### PR TITLE
scm: git: end gitignore with newline

### LINKED ISSUE #310: Newline in .gitignore

### PR DESCRIPTION
**Pull request by pared**
* [x] Have you followed the guidelines in our
      [Contributing document](https://dvc.org/doc/user-guide/contributing)?

* [x] Does your PR affect documented changes or does it add new functionality
      that should be documented? If yes, have you created a PR for
      [dvc.org](https://github.com/iterative/dvc.org) documenting it or at
      least opened an issue for it? If so, please add a link to it.

-----
Fixes #310

### CODE REVIEW

File: dvc/scm/git/__init__.py [hunk 1]
```diff
@@ -124,19 +124,16 @@ def ignore(self, path, in_curr_dir=False):
...
 
         self.track_file(os.path.relpath(gitignore))
 
         self.ignored_paths.append(path)
 
     @staticmethod
-    def _add_entry_to_gitignore(entry, gitignore, ignore_list):
-        content = entry
-        if ignore_list:
-            content = "\n" + content
```
Comment (efiop) [inline dvc/scm/git/__init__.py]: And what if there is no newline to start with? Will the current implementation just write
```
oldnew
```
?

Comment (pared) [reply dvc/scm/git/__init__.py]: Yes, fixed it and wrote test for such situation

File: dvc/scm/git/__init__.py [hunk 2]
```diff
@@ -124,19 +124,33 @@ def ignore(self, path, in_curr_dir=False):
...
+    def _append_newline_on_not_ending_with_newline(path):
+        if not os.path.exists(path):
+            return
+
+        last = None
+        with open(path, "r", encoding="utf-8") as fobj:
```
Comment (efiop) [inline dvc/scm/git/__init__.py]: Don't we check that already in https://github.com/iterative/dvc/pull/2118/files#diff-9fe1cb4148dd9740d77f52dc6d30ba5bL116 ?

Comment (pared) [reply dvc/scm/git/__init__.py]: Yes we do, but if it does not exist we don't create new one.

Comment (efiop) [reply dvc/scm/git/__init__.py]: So why are we reading this file again? And then opening it once again to add the newline, and then opening it again to write the rest. Seems suboptimal.

Comment (pared) [reply dvc/scm/git/__init__.py]: True, Ill move the logic

Comment (efiop) [reply dvc/scm/git/__init__.py]: We've already read `ignore_list` above and now we read it again to get the last line.

File: dvc/scm/git/__init__.py [hunk 3]
```diff
@@ -112,31 +124,26 @@ def ignore(self, path, in_curr_dir=False):
         )
         entry, gitignore = self._get_gitignore(path, base_dir)
 
-        ignore_list = []
-        if os.path.exists(gitignore):
-            with open(gitignore, "r") as f:
-                ignore_list = f.readlines()
-            if any(filter(lambda x: x.strip() == entry.strip(), ignore_list)):
-                return
+        ignore_list = Git._process_existing_gitignore(gitignore)
```
Comment (efiop) [inline dvc/scm/git/__init__.py]:
```suggestion
        ignore_list = self._process_existing_gitignore(gitignore)
```

File: dvc/scm/git/__init__.py [hunk 4]
```diff
@@ -112,31 +124,26 @@ def ignore(self, path, in_curr_dir=False):
...
+        if any(filter(lambda x: x.strip() == entry.strip(), ignore_list)):
```
Comment (efiop) [inline dvc/scm/git/__init__.py]: Also, how about simply
```
if self._ignored(entry, gitignore):
    return
```

File: dvc/scm/git/__init__.py [hunk 5]
```diff
@@ -112,31 +124,26 @@ def ignore(self, path, in_curr_dir=False):
...
         self.ignored_paths.append(path)
 
     @staticmethod
-    def _add_entry_to_gitignore(entry, gitignore, ignore_list):
-        content = entry
-        if ignore_list:
-            content = "\n" + content
+    def _add_entry_to_gitignore(entry, gitignore):
+
         with open(gitignore, "a", encoding="utf-8") as fobj:
```
Comment (efiop) [inline dvc/scm/git/__init__.py]: And here
```suggestion
        with open(gitignore, "a+", encoding="utf-8") as fobj:
            fobj.seek(-1, os.SEEK_END)
            prefix = "" if fobj.read(1) == "\n" else "\n"
            fobj.write("{}{}\n".format(prefix, entry))
```
?

Comment (efiop) [reply dvc/scm/git/__init__.py]: So you won't need to write anything when you are checking whether or not entry is already in gitignore above.

**Pull request merged**